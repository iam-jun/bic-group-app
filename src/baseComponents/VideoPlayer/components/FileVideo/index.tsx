import React, {
  FC, useImperativeHandle, useRef, useState,
} from 'react';
import {
  View,
  StyleSheet,
  TouchableOpacity,
  Image,
} from 'react-native';
import { Video, ResizeMode } from 'expo-av';
import { ExtendedTheme, useTheme } from '@react-navigation/native';

import { scaleCoverHeight } from '~/theme/dimension';
import Icon from '../../../Icon';
import { getVideoExtention } from '../../helper';
import LoadingIndicator from '~/beinComponents/LoadingIndicator';
import Button from '../../../Button';
import { VideoPlayerProps } from '../..';

export interface FileVideoRef {
  play: () => void,
  pause: () => void,
  resetVideoPosition: () => void,
}

const PLAYER_HEIGHT = scaleCoverHeight();

const FileVideo: FC<VideoPlayerProps> = ({
  src,
  thumbnail,
  videoRef,
  isLooping = false,
  resizeMode = ResizeMode.CONTAIN,
  ...props
}) => {
  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;
  const styles = createStyle(theme);

  const video = useRef<Video>();
  const [isPlaying, setPlaying] = useState(false);
  const [loading, setLoading] = useState(false);
  const urlExtension = getVideoExtention(src);

  useImperativeHandle(videoRef, () => ({
    play,
    pause,
    resetVideoPosition,
  }));

  const play = async () => {
    if (video.current && !!src) {
      setLoading(true);
      try {
        await video.current?.loadAsync?.({
          uri: src,
          overrideFileExtensionAndroid: urlExtension,
        });
        setPlaying(true);
        video.current?.playAsync();
        setLoading(false);
      } catch (error) {
        setLoading(false);
        console.error(
          '>>>>>>> Video loadAsync error >>>>>>>', error,
        );
      }
    }
  };

  const pause = async () => {
    const currentStatus = await video.current.getStatusAsync();
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    if (!currentStatus?.isPlaying) return;

    try {
      video.current.pauseAsync();
    } catch (error) {
      console.error(
        'STOP VIDEO FAILED>>>>>>>>>>', error,
      );
    }
  };

  const resetVideoPosition = async () => {
    if (!!video.current) {
      video.current.setStatusAsync({ positionMillis: 0, shouldPlay: false });
    }
  };

  if (!src && !thumbnail) return null;

  const renderThumbnail = () => {
    if (isPlaying) return null;

    return <Image style={styles.thumbnail} source={{ uri: thumbnail }} />;
  };

  const renderLoading = () => {
    if (!loading || isPlaying) return null;

    return <LoadingIndicator size={60} color={colors.gray20} />;
  };

  const renderPlayButton = () => {
    if (isPlaying || loading) return null;

    return (
      <Button
        TouchableComponent={TouchableOpacity}
        activeOpacity={src ? 0.85 : 1}
        style={styles.buttonPlay}
        onPress={play}
      >
        <Icon size={60} tintColor={colors.gray20} icon="CirclePlay" />
      </Button>
    );
  };

  return (
    <View style={styles.container}>
      <Video
        {...props}
        ref={video}
        style={styles.player}
        useNativeControls
        resizeMode={resizeMode}
        isLooping={isLooping}
        onError={(error: string) => {
          console.warn(
            'video failed', error,
          );
        }}
      />
      {renderThumbnail()}
      {renderLoading()}
      {renderPlayButton()}
    </View>
  );
};

const createStyle = (theme: ExtendedTheme) => {
  const { colors } = theme;
  return StyleSheet.create({
    container: {
      flex: 1,
      height: PLAYER_HEIGHT,
      justifyContent: 'center',
      backgroundColor: colors.black,
    },
    player: {
      position: 'absolute',
      width: '100%',
      height: PLAYER_HEIGHT,
    },
    thumbnail: {
      left: 0,
      top: 0,
      right: 0,
      bottom: 0,
      position: 'absolute',
      resizeMode: 'contain',
      backgroundColor: colors.black,
    },
    buttonPlay: {
      zIndex: 2,
      position: 'absolute',
      alignSelf: 'center',
    },
  });
};

export default React.forwardRef((
  props: VideoPlayerProps, ref?: React.MutableRefObject<FileVideoRef>,
) => (
  <FileVideo videoRef={ref} {...props} />
));
