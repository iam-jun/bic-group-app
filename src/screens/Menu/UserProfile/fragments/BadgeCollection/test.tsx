import React from 'react';
import {
  View, Dimensions, FlatList, StyleSheet,
} from 'react-native';
import { ExtendedTheme, useTheme } from '@react-navigation/native';
import Avatar from '~/baseComponents/Avatar';
import Icon from '~/baseComponents/Icon';
import spacing, { borderRadius } from '~/theme/spacing';

const Grid = ({ data }: {data: any[]}) => {
  const deviceWidth = Dimensions.get('window').width;

  const itemWidth = 48; // Constant item width
  const numColumns = Math.floor((deviceWidth - 10 * 2) / (itemWidth + 6 * 2));
  const rowWidth = numColumns * (itemWidth + 6 * 2);
  const paddingLeft = (deviceWidth - (10 * 2) - rowWidth) / 2;

  const theme: ExtendedTheme = useTheme();
  const { colors } = theme;
  const styles = themeStyles(theme);

  const isSelected = false;

  const renderGridItem = ({ item, index }: any) =>
    // Render each grid item here
    (
      <View
        key={index}
        style={{
          margin: 6,
          borderWidth: isSelected ? 2 : 1,
          borderColor: isSelected ? colors.purple50 : colors.neutral5,
          borderRadius: borderRadius.pill,
        }}
      >
        <Avatar.Medium
          isRounded
          source={{ uri: item?.iconUrl }}
        />
        {isSelected
          ? (
            <View style={styles.iconChose}>
              <Icon size={10} icon="Check" tintColor={colors.white} />
            </View>
          )
          : null}
      </View>
    );

  return (
    <View style={{
      flex: 1,
      justifyContent: 'center',
      alignItems: 'center',
      paddingHorizontal: 10,
      paddingBottom: 10,
    }}
    >

      <View style={{
        // paddingLeft,
      }}
      >
        <FlatList
          data={data}
          renderItem={renderGridItem}
          keyExtractor={(item, index) => index.toString()}
          numColumns={numColumns}
        />
      </View>
    </View>
  );
};

const themeStyles = (theme: ExtendedTheme) => {
  const { colors } = theme;

  return StyleSheet.create({
    container: {
      flex: 1,
      paddingTop: spacing.padding.large,
    },
    iconChose: {
      backgroundColor: colors.purple50,
      borderRadius: borderRadius.pill,
      height: 20,
      width: 20,
      alignItems: 'center',
      justifyContent: 'center',
      position: 'absolute',
      zIndex: 1,
      right: 0,
    },
  });
};

export default Grid;
