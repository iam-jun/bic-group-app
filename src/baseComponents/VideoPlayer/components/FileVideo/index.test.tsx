import * as React from 'react';
import { cleanup } from '@testing-library/react-native';

import { renderWithRedux } from '~/test/testUtils';
import FileVideo from '.';

afterEach(cleanup);

describe('VideoPlayer component', () => {
  it('renders correctly', () => {
    const rendered = renderWithRedux(
      <FileVideo src="" thumbnail="" />,
    ).toJSON();
    expect(rendered).toMatchSnapshot();
  });
});
