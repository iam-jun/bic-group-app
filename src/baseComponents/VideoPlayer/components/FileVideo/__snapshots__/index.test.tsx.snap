// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPlayer component renders correctly 1`] = `null`;
